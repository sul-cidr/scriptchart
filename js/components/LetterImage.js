import React from "react";

/* LetterImage - displays a letter image retrieved from a URL that points
 * to the image on the backend server.
 * May show 1) a binarized letter image that has been resized to a given max
 * width or height (maintaining original aspect ration), or 2) a cropped
 * manuscript image of the letter with a variably sized contextual margin
 * around it that is generated dynmaically via the backend API, or
 * 3) a binarized letter image that pops up the cropped manuscript image
 * when hovered.
 */

import Popup from "reactjs-popup";

import { API_ROOT } from "./App";

import "./index.css";

const IMAGE_DIMS = { Small: 25, Medium: 50, Large: 100 };
const CROP_MARGINS = { Small: 25, Medium: 50, Large: 100 };

class LetterImage extends React.Component {
  constructor(props) {
    super(props);

    this.resizeKeepAspect = this.resizeKeepAspect.bind(this);
    this.getCropURL = this.getCropURL.bind(this);
  }

  resizeKeepAspect(inWidth, inHeight, maxDim) {
    let hwRatio = parseFloat(inHeight) / parseFloat(inWidth);
    if (inWidth >= inHeight) {
      return { height: Math.round(hwRatio * maxDim), width: maxDim };
    } else {
      return { height: maxDim, width: Math.round(maxDim / hwRatio) };
    }
  }

  getCropURL(marginSize=0) {
    /* Compute dimensions of the cropped manuscript image, including any
     * user-specified contextual margin around it, and construct a URL
     * to request this image from the backend.
     */

    let leftExpanded = Math.max(0, this.props.coords.left - marginSize);
    let topExpanded = Math.max(0, this.props.coords.top - marginSize);
    let rightExpanded = Math.min(
      this.props.coords.pagewidth,
      this.props.coords.left + this.props.coords.width + marginSize
    );
    let widthExpanded = rightExpanded - leftExpanded;
    let bottomExpanded = Math.min(
      this.props.coords.pageheight,
      this.props.coords.top + this.props.coords.height + marginSize
    );
    let heightExpanded = bottomExpanded - topExpanded;

    if (isNaN(leftExpanded) || isNaN(topExpanded) || isNaN(widthExpanded) || isNaN(heightExpanded)) {
      console.log("NaN in getCropURL, marginSize is ")
    }

    let cropURL = API_ROOT +
      "crop?page_url=" +
      this.props.coords.pageurl +
      "&x=" +
      leftExpanded +
      "&y=" +
      topExpanded +
      "&w=" +
      widthExpanded +
      "&h=" +
      heightExpanded;

    return { url: cropURL, width: widthExpanded, height: heightExpanded };
  }

  componentDidMount() {
    /* This forces the browser to preload the image so it's ready to be
     * shown in the popup -- no state management needed!
     */
    if (this.props.showCropped) {
      const croppedImage = new Image();
      croppedImage.src = this.getCropURL().url;
    }
    let cropMargin = parseFloat(CROP_MARGINS[this.props.cropMargin]);
    let resizeRatio = (parseFloat(this.props.coords.width) * parseFloat(this.props.coords.height)) / ((cropMargin + parseFloat(this.props.coords.width)) * (cropMargin + parseFloat(this.props.coords.height)));
    let contextCropMargin = Math.round(cropMargin * resizeRatio);
    if (isNaN(contextCropMargin)) {
      console.log("contextCropMargin isNaN in componentDidMount");
      console.log(this.props.coords);
    }
    const contextImage = new Image();
    contextImage.src = this.getCropURL(contextCropMargin).url;
  }

  render() {
    let coordsWidth = this.props.coords.width;
    let coordsHeight = this.props.coords.height;

    let maxDim = IMAGE_DIMS[this.props.imageSize];
    let dims = this.resizeKeepAspect(coordsWidth, coordsHeight, maxDim);

    let binarizedDiv = '';
    let croppedDiv = '';

    // note that for the central, non-context portion of the popup image
    // to be exactly the same size as the cropped/binarized images, it
    // needs to be scaled at the same factor as the cropped/binarized
    // images themselves

    let cropMargin = parseFloat(CROP_MARGINS[this.props.cropMargin]);
    let resizeRatio = (parseFloat(coordsWidth) * parseFloat(coordsHeight)) / ((cropMargin + parseFloat(coordsWidth)) * (cropMargin + parseFloat(coordsHeight)));
    let contextCropMargin = Math.round(cropMargin * resizeRatio);
    if (isNaN(contextCropMargin)) {
      console.log("contextCropMargin isNaN in render");
    }

    let contextURL = this.getCropURL(contextCropMargin).url;

    let contextImage = <img
                         alt={this.props.letter}
                         ref="context"
                         width={dims.width + contextCropMargin}
                         height={dims.height + contextCropMargin}
                         src={contextURL}
                       />

    if (this.props.showBinarized) {

      /* Dimensions of the displayed binarized image */
      //let dims = { coordsWidth, coordsHeight };
      //let maxDim = Math.max(coordsWidth, coordsHeight);

      let binarizedImage =
        <img
          alt={this.props.letter}
          ref="binarized"
          width={dims.width}
          height={dims.height}
          src={this.props.coords.binaryurl}
        />;

      binarizedDiv =
        <div className={"letter-item"}>
          <Popup trigger={binarizedImage} position="top center" on={this.props.contextMode}>
            {contextImage}
          </Popup>
        </div>;

    }

    if (this.props.showCropped) {

      let cropURL = this.getCropURL().url;
      //let cropDims = this.resizeKeepAspect(cropData.width, cropData.height, maxDim);

      let croppedImage =
        <img
          alt={this.props.letter}
          ref="cropped"
          width={dims.width}
          height={dims.height}
          src={cropURL}
        />
        
      croppedDiv =
        <div className={"letter-item"}>
          <Popup trigger={croppedImage} position="top center" on={this.props.contextMode}>
            {contextImage}
          </Popup>
        </div>;
    }

    return (
      <span className={"letter-row"}>
        { binarizedDiv }
        { croppedDiv }
      </span>
    )
  }
}

export default LetterImage;
