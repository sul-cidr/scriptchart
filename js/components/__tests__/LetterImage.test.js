/* eslint-disable import/no-extraneous-dependencies */
/* eslint-disable no-undef */
import React from 'react';
import renderer from 'react-test-renderer';
import LetterImage from '../LetterImage';


describe('LetterImage test', () => {
  it('LetterImage should match snapshot', () => {
    const coords = 
    {
        "id": 1,
        "page": {
            "id": 1,
            "manuscript": 1,
            "url": "https://images.syriac.reclaim.hosting/manuscripts/VatSyr110/VatSyr110-002.png",
            "height": 1584,
            "width": 1044
        },
        "letter": 37,
        "top": 406,
        "left": 567,
        "width": 22,
        "height": 40,
        "binary_url": "https://images.syriac.reclaim.hosting/binary/Vat. Syr. 110_1R_Sadhe_567_406_22_40.png"
    };
    const component = renderer.create(<LetterImage coords={coords}
       />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
