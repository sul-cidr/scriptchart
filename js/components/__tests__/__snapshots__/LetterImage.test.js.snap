// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LetterImage test LetterImage should match snapshot 1`] = `
<span
  className="letter-row"
>
  
  
</span>
`;
